include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -2477,11 +3031,552 @@ namespace internal
+                  hessians_quad + n_q_points, scratch_data);
+                eval0.template values<0, false, true, normal_dir>(scratch_data,
+                                                                  values_dofs);
+
+                // grad zz
+                eval1.template values<1, false, false, normal_dir>(
+                  hessians_quad + 2 * n_q_points, scratch_data);
+                eval0.template values<0, false, false, normal_dir>(
+                  scratch_data, values_dofs + 2 * dofs_stride);
+
+                // grad xy
+                eval1.template gradients<1, false, false, normal_dir>(
+                  hessians_quad + 3 * n_q_points, scratch_data);
+                eval0.template gradients<0, false, true, normal_dir>(
+                  scratch_data, values_dofs);
+
+                // grad xz
+                eval1.template values<1, false, false, normal_dir>(
+                  hessians_quad + 4 * n_q_points, scratch_data);
+                if ((evaluation_flag & EvaluationFlags::gradients))
+                  eval0.template gradients<0, false, true, normal_dir>(
+                    scratch_data, values_dofs + dofs_stride);
+                else
+                  eval0.template gradients<0, false, false, normal_dir>(
+                    scratch_data, values_dofs + dofs_stride);
+
+                // grad yz
+                eval1.template gradients<1, false, false, normal_dir>(
+                  hessians_quad + 5 * n_q_points, scratch_data);
+                eval0.template values<0, false, true, normal_dir>(
+                  scratch_data, values_dofs + dofs_stride);
+
+                break;
+              case 2:
+                // grad xx
+                if (evaluation_flag &
+                    (EvaluationFlags::values | EvaluationFlags::gradients))
+                  eval0.template hessians<0, false, true, normal_dir>(
+                    hessians_quad, values_dofs);
+                else
+                  eval0.template hessians<0, false, false, normal_dir>(
+                    hessians_quad, values_dofs);
+
+                // grad yy
+                eval0.template values<0, false, false, normal_dir>(
+                  hessians_quad + n_q_points, values_dofs + 2 * dofs_stride);
+                // grad xy
+                if ((evaluation_flag & EvaluationFlags::gradients))
+                  eval0.template gradients<0, false, true, normal_dir>(
+                    hessians_quad + 2 * n_q_points, values_dofs + dofs_stride);
+                else
+                  eval0.template gradients<0, false, false, normal_dir>(
+                    hessians_quad + 2 * n_q_points, values_dofs + dofs_stride);
+                break;
+              default:
+                AssertThrow(false, ExcNotImplemented());
+            }
+        }
+    }
+  };
 
 
   template <int dim, int fe_degree, typename Number, bool lex_faces = false>
   struct FEFaceNormalEvaluationImpl
   {
+    using EvalNormalF =
```

Comment:
What is `F` standing for?

Reply:
>  But maybe the Piola transforms (as indicated by the todos below) will add here some additional complexity.

Yes, the Piola transform for general cells will require some changes here. 

> What is F standing for?

Just face :) I can write it out explicitly.